### PR TITLE
Reduce memory footprint for PDF generation

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
## Context
More can be done to reduce footprint (and will if we need to go further) but this ought to help.
Existing code was retrieving the all the individual PDF templates in a "configuration" as well as the generated PDF and keeping in memory the generated PDF documents.

Strategy here is to retrieve the templates one at a time and as needed, generate the corresponding document, use it and go on to next one.

This should reduce memory usage by good chunk for our multi template PDF generation. These are the biggest problems.

More can be done if needed.

Setting this up to run in dev environment. GCP profiler highlighted this functionality as a heap usage hotspot.

## Checklist

- [ X] I have labeled the type of changes involved using the `C-*` labels.
- [X ] I have assessed potential risks and labeled using the `R-*` labels.
- [ X] I have considered error handling and alerts, and added `L-*` labels as needed.
- [X ] I have considered security and privacy, and added `I-*` labels as needed
- [ X] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ X] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ X] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc
## FUD Score

_Overall, how are you feeling about these changes?_

- [X ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?
GCP profiler will show if helps.

## Testing
Existing tests cover functionality. Some refactoring was needed.
- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [X ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [X ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: .circleci/config.yml
Comment (MocanaAtBroad) [inline .circleci/config.yml:854]: Want to try out this on dev before merge.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java:150]: One of consequences of splitting PdfConfiguration and templates.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java:161]: The configuration only refers to the id. The real template object will have to be queried separately. More about this...

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java:282]: This is a rarely invoked method, so was satisfied with not writing code that would delete templates with just one operation.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java:330]: New way to retrieve templates.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/handler/EmailNotificationHandler.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/handler/EmailNotificationHandler.java:281]: Returning stream instead of bytes. Does not make a difference in terms of memory right now, but could if we do further refactoring.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/pdf/PdfConfiguration.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/pdf/PdfConfiguration.java:21]: One of the runtime memory problems trying to fix here: a configuration could have a list of templates associated with it. And each template has a reference to the the PDF blob. Objective here is to handle only one PDF blob at a time. So switching from list of templates to list of template ids.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java:151]: Retrieve on pdf template at a time for processing.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java:180]: Avoid holding the individual generated pdfs in templates in memory. Add it to working document and let it go.

File: pepper-apis/src/test/java/org/broadinstitute/ddp/service/PdfGenerationServiceTest.java
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/service/PdfGenerationServiceTest.java]: Are we sure we want to sleep during the test suite? And 30 seconds at that...

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/pdf/PdfConfiguration.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/pdf/PdfConfiguration.java]: Let's use primitive `long` to avoid possibility of `null`s.
```suggestion
    public void addTemplateId(long templateId) {
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java]: Might be better to do `orElseThrow()` and calling `deleteTemplate()` separately.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/PdfDao.sql.stg
Comment (yufengwng) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/PdfDao.sql.stg]: Doesn't look like we need the join to the `pdf_base_template` table.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java]:
```suggestion
                long templateId = configuration.getTemplateIds().get(pdfOrderIndex);
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java:174]: Do you think we should add a `pdf = null` statement here? So that we "drop" that reference at the end of this loop iteration, so that in the next loop iteration, if the JVM needs to run GC before loading in the result of `findFullTemplateByTemplateId()` then it can free up the `byte[]`? We could also simply just move the `byte[] pdf` variable declaration inside the `for` loop. That should drop that byte array reference after each iteration. Or am I totally misunderstanding things here?

**Review by yufengwng — COMMENTED** [review]
Very nice!

File: pepper-apis/src/test/java/org/broadinstitute/ddp/service/PdfGenerationServiceTest.java
Comment (MocanaAtBroad) [reply pepper-apis/src/test/java/org/broadinstitute/ddp/service/PdfGenerationServiceTest.java]: Those should not be there. Removed them (was holding up test so I could attach Profiler my test code executed).
b24e1e52

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java
Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PdfDao.java]: Sure 0892b1c0

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/PdfDao.sql.stg
Comment (MocanaAtBroad) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/PdfDao.sql.stg]: Right! Dropping join.
7bb503c8

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java:174]: Nulling it variable explicitly ought to be unnecessary. Assigning a new object to that variable and stack frame for this method getting unloaded both ought to de-reference the PDF object and make it candidate for GC reaping.
But moving the variable into the loop is a good idea. Probably marginal benefit, but definitely in the spirit of what I am trying to do in PR.
7137e2eb

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetPdfRoute.java
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetPdfRoute.java:123]: `IOUtils` is great!

File: pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/handler/EmailNotificationHandler.java
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/handler/EmailNotificationHandler.java]: Should we `close()` this stream?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfService.java
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfService.java]: Should we `close()` `pdfStream` prior to return?

**Review by zyme — DISMISSED** [review]
Overall looks good to me.  CC @Rinchan20 for some testing here both with DSM and with emailed consents.  @MocanaAtBroad I added a few notes about maybe closing some streams.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfService.java
Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfService.java]: Yes we should. Thanks! d0afd6bd

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java:506]: Changed this based on observation running unit tests locally: the font loading/creation is the most memory intensive part of running the pdf generation unit tests.
Don't think that overall this change improves memory situation: the change uses the itext font caching so it is not reloaded each time is needed. It MIGHT reduce the size of documents.

**Review by yufengwng — APPROVED** [review]
LGTM

**Pull request merged**
